yt_dlp/extractor/bilibili.py

```diff
@@ -2233,3 +2309,125 @@ def _real_extract(self, url):
                 'Referer': url,
             },
         }
+
+
+class BiliBiliSearchPageIE(BilibiliBaseIE):
+    _VALID_URL = r'https?://search\.bilibili\.com/(?P<type>all|video|bangumi|pgc|live|upuser).*'
+    _TESTS = [{
+        'url': r'https://search.bilibili.com/all?keyword=yt+-+dlp+%E4%B8%8B%E8%BD%BD%E5%99%A8',
+        'playlist_count': 20,
+        'info_dict': {
+            'id': 'yt - dlp 下载器',
+            'title': 'yt - dlp 下载器',
+        },
+    }, {
+        'url': r'https://search.bilibili.com/bangumi/?keyword=%E5%AD%A4%E7%8B%AC%E6%91%87%E6%BB%9A&from_source=webtop_search&spm_id_from=333.1007&search_source=5',
+        'playlist_mincount': 1,
+        'info_dict': {
+            'id': '孤独摇滚',
+            'title': '孤独摇滚',
+        },
+        'skip': 'geo-restricted',
+    }, {
+        'url': r'https://search.bilibili.com/video?keyword=%E8%AE%A9%E5%AD%90%E5%BC%B9%E9%A3%9E&from_source=webtop_search&spm_id_from=333.1007&search_source=5&order=dm&duration=4&tids=181&page=3&o=72',
+        'playlist_mincount': 5,
+        'info_dict': {
+            'id': '让子弹飞',
+            'title': '让子弹飞',
+        },
+    }]
+
+    def _real_extract(self, url):
+        headers = self.geo_verification_headers()
+        entries = []
+        params = parse_qs(url)
+        query = {}
+        if not self._get_cookies('https://api.bilibili.com').get('buvid3'):
+            self._set_cookie('.bilibili.com', 'buvid3', f'{uuid.uuid4()}infoc')
+        search_type = self._match_valid_url(url).group('type')
+        raw_playlist_id = traverse_obj(params, ('keyword', 0))
+        if raw_playlist_id is None:
+            raise ExtractorError('Please specify the keyword to search for!', expected=True)
+        playlist_id = urllib.parse.unquote_plus(raw_playlist_id)
+        search_type_mapping = {
+            'video': 'video',
+            'bangumi': 'media_bangumi',
+            'pgc': 'media_ft',
+            'live': 'live_room',
+            'upuser': 'bili_user',
+        }
+        valid_params = [
+            'keyword',
+            'page',
```

Comment:
Pagination varies between the webpage and the JSON API. The webpage displays 42 results per page while the JSON API returns 20 in each query.

Reply:
The metadata is misleading. The number of results are correct, but the `_TESTS` needs some update.